### ISSUE TITLE
LSTM learns but doesnt remember?

### ISSUE THREAD

**Issue by asieradzk** [dotnet/TorchSharp#1212]
Hello.

I was wondering if someone could tell me if I am making some kind of mistake when initializing my LSTM?

My network seems to be able to learn to solve non recurrent problems. Loss decreases and performance improves overtime, but when I introduce some memory-based learning it all disappears. 

For my problem the input is (sequence, numFeatures). The key information is within first 3 inputs ((1-3), numFeatures) but I can't seem to get anything useful out of LSTM module. The problem **WORKS WITH GRU**

Is there some secret to initializing it? Can someone confirm they had success with it and the problem lies somewhere else in my code?

To me it seems that LSTM either doesnt remember anything between inputs of sequences. Since its weights must be correctly updated during backpropagation (otherwise my network would also fail to learn Non-RNN problems) 

tl;dr; LSTM learns when problem doesnt require memory of pervious episodes but fails otherwise. Works fine when replaced with GRU. 

```cs
 public class PPOCriticNet1D : PPOCriticNet
 {
     private readonly ModuleList<Module<Tensor, Tensor>> fcModules = new();
     private readonly Module<Tensor, Tensor> head;
     private readonly LSTM lstmLayer;
     private int hiddenSize;
     private bool useRnn;


     public PPOCriticNet1D(string name, long inputs, int width, int depth = 3, bool useRNN = false) : base(name)
     {
         // Ensure depth is at least 1.
         if (depth < 1) throw new ArgumentOutOfRangeException("Depth must be 1 or greater.");

         this.useRnn = useRNN;
         this.hiddenSize = width;


         fcModules.Add(Linear(width, width));
         for (int i = 1; i < depth; i++)
         {
             fcModules.Add(Linear(width, width));
         }

         if (useRnn)
         {
             lstmLayer = nn.LSTM(inputs, hiddenSize, 1, batchFirst: true);

         }

         // Final layer to produce the value estimate.
         head = Linear(width, 1);

         RegisterComponents();
     }

     public override Tensor forward(Tensor x)
     {
        
         // Adjust for a single input.
         if (x.dim() == 1)
         {
             x = x.unsqueeze(0);
         }

         if (useRnn)
         {
             x = x.unsqueeze(0);
             var tuple = lstmLayer.forward(x, null);
             x = tuple.Item1.squeeze(0);
         }


         x = functional.tanh(fcModules.First().forward(x));



         foreach (var module in fcModules.Skip(1))
         {
             x = functional.tanh(module.forward(x));
         }


         var result = head.forward(x);
         return result;
     }
 


     protected override void Dispose(bool disposing)
     {
         if (disposing)
         {
 
             foreach (var module in fcModules)
             {
                 module.Dispose();
             }
             lstmLayer?.Dispose();
             head.Dispose();
         }

         base.Dispose(disposing);
     }
 }
 ```
